### PR TITLE
(#9302) Add Ruby 3.3 to CI matrix

### PR DESCRIPTION
**Pull request by bastelfreak**

### CODE REVIEW

**Comment by mhashizume** [comment]
Hi @bastelfreak , now that https://github.com/puppetlabs/puppet/pull/9398 has been merged, can you rebase your PR? Thanks!

**Comment by bastelfreak** [comment]
@mhashizume rebased it

**Comment by joshcooper** [comment]
@bastelfreak It seems ruby 3.3 changed the error message again, could you relax these expectations:
https://github.com/puppetlabs/puppet/blob/fa0e42a70034cbd27ed4884f6b313e33c43b84a4/spec/unit/x509/cert_provider_spec.rb#L283

https://github.com/puppetlabs/puppet/blob/fa0e42a70034cbd27ed4884f6b313e33c43b84a4/spec/unit/x509/cert_provider_spec.rb#L318

**Comment by bastelfreak** [comment]
mhm I'm not sure why the windows runner doesn't find any spec files:

```
error: no rspec examples were found.
rake aborted!
error: no specs were found.
D:/a/puppet/puppet/rakelib/parallel.rake:214:in `group_specs'
D:/a/puppet/puppet/rakelib/parallel.rake:183:in `run'
D:/a/puppet/puppet/rakelib/parallel.rake:407:in `block (2 levels) in <top (required)>'
D:/a/puppet/puppet/vendor/bundle/ruby/3.3.0/gems/rake-13.2.1/exe/rake:27:in `<top (required)>'
Tasks: TOP => parallel:spec
(See full trace by running task with --trace)
```

**Comment by joshcooper** [comment]
I was able to reproduce using:

```
C:\Users\Administrator\puppet>bundle exec ruby util/rspec_grouper 1000

An error occurred while loading ./spec/integration/agent/logging_spec.rb. - Did you mean?
                    rspec ./spec/integration/util/settings_spec.rb
                    rspec ./spec/integration/type/notify_spec.rb
                    rspec ./spec/integration/data_binding_spec.rb

Failure/Error: require 'racc/parser.rb'

LoadError:
  cannot load such file -- racc/parser.rb
# ./lib/puppet/pops/parser/eparser.rb:7:in `<top (required)>'
...
```

To resolve this issue for right now, I suggest changing this line in `Gemfile`

```
C:\Users\Administrator\puppet>git diff
diff --git a/Gemfile b/Gemfile
index 5d3caa47a3..ec0cb70688 100644
--- a/Gemfile
+++ b/Gemfile
@@ -62,7 +62,7 @@ end
 group(:development, optional: true) do
   gem 'memory_profiler', require: false, platforms: [:mri]
   gem 'pry', require: false, platforms: [:ruby]
-  gem "racc", "1.5.2", require: false, platforms: [:ruby]
+  gem "racc", "1.5.2", require: false
   if RUBY_PLATFORM != 'java'
     gem 'ruby-prof', '>= 0.16.0', require: false
   end

C:\Users\Administrator\puppet>bundle update
Fetching gem metadata from https://rubygems.org/.........
Resolving dependencies...
Bundle updated!

C:\Users\Administrator\puppet>bundle exec ruby -e 'require "racc"; puts Racc::VERSION'
1.5.2

C:\Users\Administrator\puppet>bundle exec ruby util/rspec_grouper 1000
Grouped 27030 rspec example(s) into 23 group(s) from 746 file(s).

C:/Users/Administrator/AppData/Local/Temp/1/rspecrun20240912-5408-l8m3b4/d20240912-5408-jix1k5/group1
C:/Users/Administrator/AppData/Local/Temp/1/rspecrun20240912-5408-l8m3b4/d20240912-5408-jix1k5/group2
C:/Users/Administrator/AppData/Local/Temp/1/rspecrun20240912-5408-l8m3b4/d20240912-5408-jix1k5/group3
...
```

I'm going to file a separate issue about fixing our racc dependency. The problem is we use `racc 1.5.2` to generate our `eparser.rb`. But the resulting parser has a runtime dependency on `racc` https://github.com/puppetlabs/puppet/blob/7cc46f776d836fe5d07148f3ab6865f4f9e05bbc/lib/puppet/pops/parser/eparser.rb#L7

At the same time, Ruby 3.3 switched `racc` from a `default` gem to a `bundled` gem https://stdgems.org/racc/ So there's no guarantee that `racc` will be available when running in a bundler context.

We don't want to add `racc` as a hard runtime requirement, because it contains native extensions (so you'd need a compiler to gem install puppet).

I think the right thing is to use `racc -E` so that it generates a standalone `eparser.rb` which has no runtime dependency on `racc`.

**Comment by bastelfreak** [comment]
Added it, lets see.

**Comment by bastelfreak** [comment]
mhm wnidows still fails :thinking:

File: Gemfile [hunk 1]
```diff
@@ -62,7 +62,7 @@ end
 group(:development, optional: true) do
   gem 'memory_profiler', require: false, platforms: [:mri]
   gem 'pry', require: false, platforms: [:ruby]
-  gem "racc", "1.5.2", require: false, platforms: [:ruby]
+  gem "racc", "1.5.2", require: false
```
Comment (joshcooper) [inline Gemfile:65]: Can you try moving this dependency to the `test` group (and leave the `platforms: [:ruby]` omitted)?

**Review by joshcooper — APPROVED** [review]

**Comment by joshcooper** [comment]
Thanks @bastelfreak!

**Pull request merged**